In [ ]:
! pip install plotly --upgrade

     |████████████████████████████████| 21.8 MB 1.5 MB/s 
  Attempting uninstall: plotly
    Found existing installation: plotly 4.4.1
    Uninstalling plotly-4.4.1:
      Successfully uninstalled plotly-4.4.1


In [ ]:
! pip install PyMySQL

     |████████████████████████████████| 43 kB 937 kB/s 


In [ ]:
# 강의자료대로 했을 때 오류난다면 이 부분 추가!

! python3 -m pip install PyMySQL[rsa]

     |████████████████████████████████| 3.0 MB 4.1 MB/s 


In [ ]:
# 강의자료대로 했을 때 오류난다면 이 부분 추가!
# 오류나는 건 괜찮음

! pip3 install crytography

ERROR: Could not find a version that satisfies the requirement crytography (from versions: 0.0.1, 1.0.0)
ERROR: No matching distribution found for crytography


In [ ]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import matplotlib.pyplot as plt
import plotly.express as px
#from pymysql import ProgrammingError

In [ ]:
plt.style.use('seaborn')

In [ ]:
import pymysql.cursors

In [ ]:
connection = pymysql.connect(host='3.38.95.162',
                             user='USER_SQLSTK',
                             password='1qaz2wsx',
                             database='DB_SQLSTK',
                             charset='utf8',
                             cursorclass=pymysql.cursors.DictCursor
                             )

In [ ]:
connection

In [ ]:
cursor = connection.cursor()

In [ ]:
cursor

In [ ]:
sql = ''
sql = '''
SELECT T1.STK_CD, T1.DT
      , T1.C_PRC, T1.O_PRC, T1.H_PRC, T1.L_PRC, T1.VOL
FROM history_dt T1
WHERE T1.STK_CD = 121800
AND T1.DT >= '20190101'
AND T1.DT < '20191231'
;
'''

In [ ]:
cursor.execute(sql)
res = cursor.fetchall()
res

[{'C_PRC': Decimal('7246.000'),
  'DT': datetime.date(2019, 1, 2),
  'H_PRC': Decimal('7446.000'),
  'L_PRC': Decimal('7182.000'),
  'O_PRC': Decimal('7318.000'),
  'STK_CD': '121800',
  'VOL': Decimal('142598.000')},
 {'C_PRC': Decimal('7191.000'),
  'DT': datetime.date(2019, 1, 3),
  'H_PRC': Decimal('7418.000'),
  'L_PRC': Decimal('7118.000'),
  'O_PRC': Decimal('7373.000'),
  'STK_CD': '121800',
  'VOL': Decimal('172781.000')},
 {'C_PRC': Decimal('6909.000'),
  'DT': datetime.date(2019, 1, 4),
  'H_PRC': Decimal('7091.000'),
  'L_PRC': Decimal('6627.000'),
  'O_PRC': Decimal('7091.000'),
  'STK_CD': '121800',
  'VOL': Decimal('370164.000')},
 {'C_PRC': Decimal('6800.000'),
  'DT': datetime.date(2019, 1, 7),
  'H_PRC': Decimal('7055.000'),
  'L_PRC': Decimal('6718.000'),
  'O_PRC': Decimal('7046.000'),
  'STK_CD': '121800',
  'VOL': Decimal('299267.000')},
 {'C_PRC': Decimal('6764.000'),
  'DT': datetime.date(2019, 1, 8),
  'H_PRC': Decimal('6909.000'),
  'L_PRC': Decimal('6700.000'

In [ ]:
dt_df = pd.DataFrame(res)
dt_df

STK_CD          DT     C_PRC     O_PRC     H_PRC     L_PRC          VOL
0    121800  2019-01-02  7246.000  7318.000  7446.000  7182.000   142598.000
1    121800  2019-01-03  7191.000  7373.000  7418.000  7118.000   172781.000
2    121800  2019-01-04  6909.000  7091.000  7091.000  6627.000   370164.000
3    121800  2019-01-07  6800.000  7046.000  7055.000  6718.000   299267.000
4    121800  2019-01-08  6764.000  6809.000  6909.000  6700.000   196713.000
..      ...         ...       ...       ...       ...       ...          ...
240  121800  2019-12-23  7840.000  8050.000  8060.000  7820.000   342403.000
241  121800  2019-12-24  7730.000  7820.000  7910.000  7710.000   369715.000
242  121800  2019-12-26  7800.000  7710.000  7910.000  7620.000   360120.000
243  121800  2019-12-27  7990.000  7850.000  8070.000  7800.000   397518.000
244  121800  2019-12-30  7730.000  7620.000  7730.000  7200.000  1161649.000

[245 rows x 7 columns]

In [ ]:
dt_df['DT'] = pd.to_datetime(dt_df['DT'])

In [ ]:
dt_df.loc[:, 'C_PRC' : ] = dt_df.loc[:, 'C_PRC' : ].astype(int)

In [ ]:
dt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245 entries, 0 to 244
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   STK_CD  245 non-null    object        
 1   DT      245 non-null    datetime64[ns]
 2   C_PRC   245 non-null    int64         
 3   O_PRC   245 non-null    int64         
 4   H_PRC   245 non-null    int64         
 5   L_PRC   245 non-null    int64         
 6   VOL     245 non-null    int64         
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 13.5+ KB


In [ ]:
dt_df['DIFF'] = dt_df['C_PRC'].diff()

In [ ]:
def UD(diff):
  ups = 0
  downs = 0
  if diff > 0:
    ups = diff
  elif diff < 0:
    downs = abs(diff)
  return pd.Series([ups, downs])

In [ ]:
dt_df[['U', 'D']] = dt_df['DIFF'].apply(UD)

In [ ]:
dt_df['AU'] = dt_df['U'].rolling(14).mean()

In [ ]:
dt_df['AD'] = dt_df['D'].rolling(14).mean()

In [ ]:
dt_df['RS'] = dt_df['AU'] / dt_df['AD']

In [ ]:
dt_df['RSI'] = dt_df['RS'] / (1 + dt_df['RS'])

In [ ]:
dt_df['RSI_signal'] = dt_df['RSI'].rolling(6).mean()

In [ ]:
dt_df = dt_df.set_index('DT')

In [ ]:
dt_df['S3'] = 0.3
dt_df['S7'] = 0.7

In [ ]:
dt_df.reset_index(inplace=True)

매수일 산정

In [ ]:
def TimeToBuy(e):
  timing = 0
  if e['RSI'] < 0.3 and e['RSI_signal'] < 0.3:
    timing = 1
  return timing

In [ ]:
dt_df['time_to_buy'] = dt_df[['RSI', 'RSI_signal']].apply(TimeToBuy, axis=1)

In [ ]:
pd.set_option('display.max_rows', None)

In [ ]:
pd.set_option('display.max_columns', None)

In [ ]:
dt_df['first_crashed'] = dt_df['time_to_buy'].apply('diff')

In [ ]:
idx = 0
idx_ls = []
fs_np = np.array(dt_df['first_crashed'])

idx = np.where(fs_np == 1.0)
idx_ls.append(idx[0]+1)

In [ ]:
idx

(array([ 24,  81, 138, 142, 228, 241]),)

In [ ]:
idx_ls

[array([ 25,  82, 139, 143, 229, 242])]

In [ ]:
purchase_date = []
for idx in idx_ls:
  purchase_date.append(dt_df.iloc[idx]['DT'])

In [ ]:
# 매수일 결정!
purchase_date

[25    2019-02-11
 82    2019-05-03
 139   2019-07-25
 143   2019-07-31
 229   2019-12-06
 242   2019-12-26
 Name: DT, dtype: datetime64[ns]]

In [ ]:
#dt_df

매도일 산정

def TimeToSell(e): # e는 Ser
  timing = 0
  if e['RSI'] > 0.7 and e['RSI_signal'] > 0.7:
    timing = 1
  return timing


dt_df['time_to_sell'] = dt_df[['RSI', 'RSI_signal']].apply('TimeToSell', axis=1)

-----------------------------------------------------------------------------
.diff(2) = 1

dt_df['first_rasing'] = dt_df['time_to_sell'].apply('diff')

idx = 0
idx_ls = []
fs_np = np.array(dt_df['first_rasing'])

idx = np.where(fs_np == 1.0)
idx_ls.append(idx[0]+1)

idx_ls에 있는 인덱스들의 dt컬럼에 있는 날짜가 매도일이 된다.
매도일 알아보기
sale_date = []
for idx in idx_ls:
  sale_date.append(dt_df.iloc[idx]['dt'])

In [ ]:
def TimeToSell(e): # e는 Ser
  timing = 0
  if e['RSI'] > 0.7 and e['RSI_signal'] > 0.7:
    timing = 1
  return timing

In [ ]:
dt_df['time_to_sell'] = dt_df[['RSI', 'RSI_signal']].apply(TimeToSell, axis=1)

In [ ]:
dt_df['first_rasing'] = dt_df['time_to_sell'].apply('diff')

In [ ]:
idx = 0
idx_ls = []
fs_np = np.array(dt_df['first_rasing'])

idx = np.where(fs_np == 1.0)
idx_ls.append(idx[0]+1)

In [ ]:
idx

(array([119, 177, 205, 207]),)

In [ ]:
idx_ls

[array([120, 178, 206, 208])]

In [ ]:
sale_date = []
for idx in idx_ls:
  sale_date.append(dt_df.iloc[idx]['DT'])

In [ ]:
# 매도일 결정!
sale_date

[120   2019-06-28
 178   2019-09-23
 206   2019-11-05
 208   2019-11-07
 Name: DT, dtype: datetime64[ns]]

dt_df[ dt_df['dt'] == '2019-12-09' ]['RSI']



dt_df['invest_share']

In [ ]:
dt_df

DT  STK_CD  C_PRC  O_PRC  H_PRC  L_PRC       VOL    DIFF       U  \
0   2019-01-02  121800   7246   7318   7446   7182    142598     NaN     0.0   
1   2019-01-03  121800   7191   7373   7418   7118    172781   -55.0     0.0   
2   2019-01-04  121800   6909   7091   7091   6627    370164  -282.0     0.0   
3   2019-01-07  121800   6800   7046   7055   6718    299267  -109.0     0.0   
4   2019-01-08  121800   6764   6809   6909   6700    196713   -36.0     0.0   
5   2019-01-09  121800   7073   6727   7109   6727    279987   309.0   309.0   
6   2019-01-10  121800   7227   7136   7318   7036    197955   154.0   154.0   
7   2019-01-11  121800   7100   7073   7264   7036    242986  -127.0     0.0   
8   2019-01-14  121800   7000   6846   7173   6746    336558  -100.0     0.0   
9   2019-01-15  121800   7350   7140   7490   7130    248755   350.0   350.0   
10  2019-01-16  121800   7250   7350   7380   7190    118368  -100.0     0.0   
11  2019-01-17  121800   7170   7230   7460   7110    253037   -80.0     0.0   
12  2019-01-18  121800   7020   7150   7180   6940    333930  -150.0     0.0   
13  2019-01-21  121800   6730   7000   7000   6610    523895  -290.0     0.0   
14  2019-01-22  121800   6700   6750   6830   6670    210028   -30.0     0.0   
15  2019-01-23  121800   6690   6670   6820   6640    183417   -10.0     0.0   
16  2019-01-24  121800   6740   6770   6770   6630    168824    50.0    50.0   
17  2019-01-25  121800   6960   6750   7070   6700    388018   220.0   220.0   
18  2019-01-28  121800   6930   6950   6970   6820    217665   -30.0     0.0   
19  2019-01-29  121800   6700   6920   6920   6670    239367  -230.0     0.0   
20  2019-01-30  121800   6560   6740   6740   6450    378649  -140.0     0.0   
21  2019-01-31  121800   6600   6570   6630   6510    164373    40.0    40.0   
22  2019-02-01  121800   6150   6520   6550   6010    673459  -450.0     0.0   
23  2019-02-07  121800   5620   6090   6090   5500   1024430  -530.0     0.0   
24  2019-02-08  121800   5710   5640   5810   5600    365274    90.0    90.0   
25  2019-02-11  121800   5750   5870   5940   5740    418093    40.0    40.0   
26  2019-02-12  121800   5880   5720   6200   5600   1278288   130.0   130.0   
27  2019-02-13  121800   5970   5900   6080   5830    385251    90.0    90.0   
28  2019-02-14  121800   6280   6240   6440   6110   1075251   310.0   310.0   
29  2019-02-15  121800   6330   6340   6380   6090    524383    50.0    50.0   
30  2019-02-18  121800   6500   6350   6560   6310    524104   170.0   170.0   
31  2019-02-19  121800   6450   6640   6700   6440    377379   -50.0     0.0   
32  2019-02-20  121800   6480   6410   6560   6380    278262    30.0    30.0   
33  2019-02-21  121800   6820   6530   6980   6490    931808   340.0   340.0   
34  2019-02-22  121800   6700   6820   6820   6680    326295  -120.0     0.0   
35  2019-02-25  121800   6710   6660   6800   6510    222065    10.0    10.0   
36  2019-02-26  121800   6470   6770   6780   6470    246013  -240.0     0.0   
37  2019-02-27  121800   6600   6450   6710   6450    250977   130.0   130.0   
38  2019-02-28  121800   6200   6660   6660   6160    372271  -400.0     0.0   
39  2019-03-04  121800   6330   6250   6470   6230    193296   130.0   130.0   
40  2019-03-05  121800   6250   6310   6410   6130    214772   -80.0     0.0   
41  2019-03-06  121800   6600   6300   6690   6300    700400   350.0   350.0   
42  2019-03-07  121800   6920   6590   7060   6590   1160719   320.0   320.0   
43  2019-03-08  121800   6890   6940   7100   6750    470647   -30.0     0.0   
44  2019-03-11  121800   6530   6790   6850   6460    435699  -360.0     0.0   
45  2019-03-12  121800   6540   6520   6670   6420    280104    10.0    10.0   
46  2019-03-13  121800   6380   6550   6550   6320    271024  -160.0     0.0   
47  2019-03-14  121800   6670   6450   6670   6390    271159   290.0   290.0   
48  2019-03-15  121800   6350   6670   6790   6330    512126  -320.0     0.0   
49  2019-03-


매수일
25    2019-02-11 -> 매수!
82    2019-05-03
139   2019-07-25 -> 2차 매수!
143   2019-07-31
229   2019-12-06
242   2019-12-26

총 매수 주식 수 :  (주)
매수를 마치고 남은 잔액 :  원

매수
2/11
원 어치 매수
25    2019-02-11

2/11 종가를 기준으로 주식수량 산정
2019/02/11 종가 : 5,750 원
약 1억원으로 17,391 주 매수가능

매수에  원 사용
잔액 :  원

In [ ]:
# 2019/02/11 종가 : 5,750 원
dt_df['C_PRC'][25]

5750

In [ ]:
# 약 1억원으로 17,391 주 매수가능
int(100000000 / dt_df['C_PRC'][25])

17391

In [ ]:
# 매수에 99,998,250 원 사용
# 잔액 : 1,750 원
dt_df['C_PRC'][25] * 17391

99998250

In [ ]:
# 매수 후 잔액
100000000 - 99998250

1750

매도하기

매수를 마친 잔액은 1,750 원
17,391 주 보유중

120   2019-06-28 -> 매도!
178   2019-09-23 -> 2차 매도!
206   2019-11-05
208   2019-11-07

In [ ]:
# 매도일인 2019-06-28에 종가 : 8,900원
dt_df['C_PRC'][120]

8900

매도 이전 총 보유 주식 수 : 17,391 주

매도 가능한 기준이 되는 첫번째 시점에 전량 매도!
120   2019-06-28 (17,391 주 매도)
매도일에 종가 : 8,900 원
매도 수익금 =  154,779,900 원

최종잔액 =  154,781,650 원

In [ ]:
# 매도 수익금 = 154,779,900 원
dt_df['C_PRC'][120] * 17391

154779900

In [ ]:
# 순 수익금 : 54,781,650 원
154779900 - 99998250

54781650

In [ ]:
# 최종잔액 = 154,781,650 원
1750 + 154779900

154781650

In [ ]:
# 1차!!!
# BUY(산 날) 종가 : 5,750(원)
buy_c_prc = 5750

In [ ]:
# SELL(판 날) 종가 : 8,900(원)
sell_c_prc = 8900

In [ ]:
# 1차 수익률
( (sell_c_prc - buy_c_prc) / buy_c_prc ) * 100

54.78260869565217

# 2차 매수/매도

매수일
25    2019-02-11 -> 매수!
82    2019-05-03
139   2019-07-25 -> 2차 매수!
143   2019-07-31
229   2019-12-06
242   2019-12-26

총 매수 주식 수 :  (주)
매수를 마치고 남은 잔액 :  원

매수
7/25
원 어치 매수
139   2019-07-25

7/25 종가를 기준으로 주식수량 산정
2019/07/25 종가 :  원
잔액( 154,781,650 원)으로  주 매수가능

매수에  원 사용
잔액 :  원

In [ ]:
# 2019/07/25 종가 : 7,020 원
dt_df['C_PRC'][139]

7020

In [ ]:
# 154,781,650 원으로 22,048 주 매수가능
int(154781650 / dt_df['C_PRC'][139])

22048

In [ ]:
# 매수에 154,776,960 원 사용
# 잔액 : 1,750 원
dt_df['C_PRC'][139] * 22048

154776960

In [ ]:
# 매수 후 잔액 : 4,690 원
154781650 - 154776960

4690

매도하기

매수를 마친 잔액은 4,690 원
22,048 주 보유중

120   2019-06-28 -> 매도!
178   2019-09-23 -> 2차 매도!
206   2019-11-05
208   2019-11-07

In [ ]:
# 매도일인 2019-09-23에 종가 : 9,480원
dt_df['C_PRC'][178]

9480

매도 이전 총 보유 주식 수 : 22,048 주

매도 가능한 기준이 되는 첫번째 시점에 전량 매도!
178   2019-09-23 (22,048 주 매도)
매도일에 종가 : 9,480 원
매도 수익금 =   209,015,040 원

최종잔액 =   원

In [ ]:
# 매도 수익금 = 209,015,040 원
dt_df['C_PRC'][178] * 22048

209015040

In [ ]:
# 순 수익금 : 54,238,080 원
209015040 - 154776960

54238080

In [ ]:
# 최종잔액 = 209,019,730 원
4690 + 209015040

209019730

In [ ]:
# 2차!!!
# BUY(산 날) 종가 : 7,020(원)
buy_c_prc = 7020

In [ ]:
# SELL(판 날) 종가 : 9,480(원)
sell_c_prc = 9480

In [ ]:
# 2차 수익률
( (sell_c_prc - buy_c_prc) / buy_c_prc ) * 100

35.04273504273504

In [ ]:
fig = make_subplots(
    rows=4, cols=1,
    vertical_spacing=0.01,
    shared_xaxes=True,
    subplot_titles=('<b>비덴트 봉차트</b>',
                    '<b>Select Data Range</b>',
                    '<b>RSI & RSI signal</b>')
)

# 봉차트
fig.add_trace(
    go.Candlestick(
        x = dt_df['DT'],
        open = dt_df['O_PRC'],
        high = dt_df['H_PRC'],
        low = dt_df['L_PRC'],
        close = dt_df['C_PRC'],
        increasing_line_color = 'red',
        decreasing_line_color = 'blue'
    ), row=1, col=1
)

# RSI 선 그래프
fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['RSI'],
        line = dict(color='blue'),
        name='RSI'
    ), row=3, col=1

)

# RSI signal 선 그래프
fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['RSI_signal'],
        line = dict(color='red')
        , name='RSI_signal'
    ), row=3, col=1

)

fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['S3'],
        line = dict(color='red', dash='dash')
        , name='High bound'
    ), row=3, col=1
)

fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['S7'],
        line = dict(color='blue', dash='dash')
        , name='Low bound'
    ), row=3, col=1
)

# 매수 적합시점 표시(실제 매수날 제외)
fig.add_trace(
    go.Scatter(
        x = ['2019-05-03', '2019-07-31', '2019-12-06', '2019-12-26'],
        y = [0.357143, 0.253968, 0.182099, 0.297143],
        mode = "markers+text",
        marker = dict(color = 'darkred', size = 10, symbol = 6),
        text = ['05-03', '07-31', '12-06', '12-26'],
        textfont = dict(size = 15),
        textposition = "bottom center",
        name = 'RSI 기반 매수 적합시점'
    ), row = 3, col = 1
)

# 실제 매수 시점 표시
fig.add_trace(
    go.Scatter(
        x = ['2019-02-11', '2019-07-25'],
        y = [0.191304, 0.340361],
        mode = "markers+text",
        marker = dict(color = 'black', size = 10, symbol = 17),
        text = ["02-11", '07-25'],
        textfont = dict(size = 15),
        textposition = "bottom center",
        name = 'RSI 기반 매수시점'
    ), row = 3, col = 1
)

# 매도 적합시점 표시(실제 매도시점 제외)
fig.add_trace(
    go.Scatter(
        x = ['2019-11-05', '2019-11-07'],
        y = [0.679812, 0.652066],
        mode = "markers+text",
        marker = dict(color = 'darkcyan', size = 10, symbol = 5),
        text = ["11-05", '11-07'],
        textfont = dict(size = 13),
        textposition = "top center",
        name = 'RSI 기반 매도 적합시점'
    ), row = 3, col = 1
)

# 실제 매도 시점 표시
fig.add_trace(
    go.Scatter(
        x = ['2019-06-28', '2019-09-23'],
        y = [0.746141, 0.821622],
        mode = "markers+text",
        marker = dict(color = 'black', size = 10, symbol = 17),
        text = ["06-28", '09-23'],
        textfont = dict(size = 13),
        textposition = "top center",
        name = 'RSI 기반 매도시점'
    ), row = 3, col = 1
)

fig.update_layout(width = 1200, height = 800, xaxis_rangeslider_visible = True)

fig.show()

# 바른손이앤에이

In [ ]:
sql = ''
sql = '''
SELECT T1.STK_CD, T1.DT
      , T1.C_PRC, T1.O_PRC, T1.H_PRC, T1.L_PRC, T1.VOL
FROM history_dt T1
WHERE T1.STK_CD = 035620
AND T1.DT >= '20190101'
AND T1.DT < '20191231'
;
'''

In [ ]:
cursor.execute(sql)
res = cursor.fetchall()
res

[{'C_PRC': Decimal('1260.000'),
  'DT': datetime.date(2019, 1, 2),
  'H_PRC': Decimal('1270.000'),
  'L_PRC': Decimal('1220.000'),
  'O_PRC': Decimal('1225.000'),
  'STK_CD': '035620',
  'VOL': Decimal('108569.000')},
 {'C_PRC': Decimal('1315.000'),
  'DT': datetime.date(2019, 1, 3),
  'H_PRC': Decimal('1415.000'),
  'L_PRC': Decimal('1265.000'),
  'O_PRC': Decimal('1265.000'),
  'STK_CD': '035620',
  'VOL': Decimal('1565090.000')},
 {'C_PRC': Decimal('1340.000'),
  'DT': datetime.date(2019, 1, 4),
  'H_PRC': Decimal('1450.000'),
  'L_PRC': Decimal('1255.000'),
  'O_PRC': Decimal('1320.000'),
  'STK_CD': '035620',
  'VOL': Decimal('1691268.000')},
 {'C_PRC': Decimal('1315.000'),
  'DT': datetime.date(2019, 1, 7),
  'H_PRC': Decimal('1370.000'),
  'L_PRC': Decimal('1310.000'),
  'O_PRC': Decimal('1340.000'),
  'STK_CD': '035620',
  'VOL': Decimal('300689.000')},
 {'C_PRC': Decimal('1300.000'),
  'DT': datetime.date(2019, 1, 8),
  'H_PRC': Decimal('1320.000'),
  'L_PRC': Decimal('1240.00

In [ ]:
dt_df = pd.DataFrame(res)
dt_df

STK_CD          DT     C_PRC     O_PRC     H_PRC     L_PRC          VOL
0    035620  2019-01-02  1260.000  1225.000  1270.000  1220.000   108569.000
1    035620  2019-01-03  1315.000  1265.000  1415.000  1265.000  1565090.000
2    035620  2019-01-04  1340.000  1320.000  1450.000  1255.000  1691268.000
3    035620  2019-01-07  1315.000  1340.000  1370.000  1310.000   300689.000
4    035620  2019-01-08  1300.000  1295.000  1320.000  1240.000   396085.000
..      ...         ...       ...       ...       ...       ...          ...
241  035620  2019-12-23  1580.000  1640.000  1640.000  1565.000   442654.000
242  035620  2019-12-24  1575.000  1580.000  1590.000  1550.000   351206.000
243  035620  2019-12-26  1565.000  1565.000  1585.000  1560.000   445074.000
244  035620  2019-12-27  1610.000  1565.000  1625.000  1560.000   770924.000
245  035620  2019-12-30  1650.000  1620.000  1660.000  1600.000   818099.000

[246 rows x 7 columns]

In [ ]:
dt_df['DT'] = pd.to_datetime(dt_df['DT'])

In [ ]:
dt_df.loc[:, 'C_PRC' : ] = dt_df.loc[:, 'C_PRC' : ].astype(int)

In [ ]:
dt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246 entries, 0 to 245
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   STK_CD  246 non-null    object        
 1   DT      246 non-null    datetime64[ns]
 2   C_PRC   246 non-null    int64         
 3   O_PRC   246 non-null    int64         
 4   H_PRC   246 non-null    int64         
 5   L_PRC   246 non-null    int64         
 6   VOL     246 non-null    int64         
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 13.6+ KB


In [ ]:
dt_df['DIFF'] = dt_df['C_PRC'].diff()

In [ ]:
dt_df

STK_CD         DT  C_PRC  O_PRC  H_PRC  L_PRC      VOL  DIFF
0    035620 2019-01-02   1260   1225   1270   1220   108569   NaN
1    035620 2019-01-03   1315   1265   1415   1265  1565090  55.0
2    035620 2019-01-04   1340   1320   1450   1255  1691268  25.0
3    035620 2019-01-07   1315   1340   1370   1310   300689 -25.0
4    035620 2019-01-08   1300   1295   1320   1240   396085 -15.0
..      ...        ...    ...    ...    ...    ...      ...   ...
241  035620 2019-12-23   1580   1640   1640   1565   442654   0.0
242  035620 2019-12-24   1575   1580   1590   1550   351206  -5.0
243  035620 2019-12-26   1565   1565   1585   1560   445074 -10.0
244  035620 2019-12-27   1610   1565   1625   1560   770924  45.0
245  035620 2019-12-30   1650   1620   1660   1600   818099  40.0

[246 rows x 8 columns]

In [ ]:
def UD(diff):
  ups = 0
  downs = 0
  if diff > 0:
    ups = diff
  elif diff < 0:
    downs = abs(diff)
  return pd.Series([ups, downs])

In [ ]:
dt_df[['U', 'D']] = dt_df['DIFF'].apply(UD)

In [ ]:
dt_df

STK_CD         DT  C_PRC  O_PRC  H_PRC  L_PRC      VOL  DIFF     U     D
0    035620 2019-01-02   1260   1225   1270   1220   108569   NaN   0.0   0.0
1    035620 2019-01-03   1315   1265   1415   1265  1565090  55.0  55.0   0.0
2    035620 2019-01-04   1340   1320   1450   1255  1691268  25.0  25.0   0.0
3    035620 2019-01-07   1315   1340   1370   1310   300689 -25.0   0.0  25.0
4    035620 2019-01-08   1300   1295   1320   1240   396085 -15.0   0.0  15.0
..      ...        ...    ...    ...    ...    ...      ...   ...   ...   ...
241  035620 2019-12-23   1580   1640   1640   1565   442654   0.0   0.0   0.0
242  035620 2019-12-24   1575   1580   1590   1550   351206  -5.0   0.0   5.0
243  035620 2019-12-26   1565   1565   1585   1560   445074 -10.0   0.0  10.0
244  035620 2019-12-27   1610   1565   1625   1560   770924  45.0  45.0   0.0
245  035620 2019-12-30   1650   1620   1660   1600   818099  40.0  40.0   0.0

[246 rows x 10 columns]

In [ ]:
dt_df['AU'] = dt_df['U'].rolling(14).mean()

In [ ]:
dt_df['AD'] = dt_df['D'].rolling(14).mean()

In [ ]:
dt_df.tail()

STK_CD         DT  C_PRC  O_PRC  ...     U     D         AU        AD
241  035620 2019-12-23   1580   1640  ...   0.0   0.0  15.357143  4.642857
242  035620 2019-12-24   1575   1580  ...   0.0   5.0  15.357143  4.642857
243  035620 2019-12-26   1565   1565  ...   0.0  10.0  15.357143  2.857143
244  035620 2019-12-27   1610   1565  ...  45.0   0.0  17.142857  2.857143
245  035620 2019-12-30   1650   1620  ...  40.0   0.0  17.857143  2.857143

[5 rows x 12 columns]

In [ ]:
dt_df['RS'] = dt_df['AU'] / dt_df['AD']

In [ ]:
dt_df['RSI'] = dt_df['RS'] / (1 + dt_df['RS'])

In [ ]:
dt_df['RSI_signal'] = dt_df['RSI'].rolling(6).mean()

In [ ]:
dt_df = dt_df.set_index('DT')

In [ ]:
dt_df['S3'] = 0.3
dt_df['S7'] = 0.7

In [ ]:
dt_df.tail()

STK_CD  C_PRC  O_PRC  H_PRC  ...       RSI  RSI_signal   S3   S7
DT                                       ...                                
2019-12-23  035620   1580   1640   1640  ...  0.767857    0.678321  0.3  0.7
2019-12-24  035620   1575   1580   1590  ...  0.767857    0.696003  0.3  0.7
2019-12-26  035620   1565   1565   1585  ...  0.843137    0.731764  0.3  0.7
2019-12-27  035620   1610   1565   1625  ...  0.857143    0.770756  0.3  0.7
2019-12-30  035620   1650   1620   1660  ...  0.862069    0.802455  0.3  0.7

[5 rows x 16 columns]

In [ ]:
dt_df.reset_index(inplace=True)

매수일 산정

def TimeToBuy(e):
  timing = 0
  if e['RSI'] < 0.3 and e['RSI_signal'] < 0.3:
    timing = 1
  return timing


dt_df['time_to_buy'] = dt_df[['RSI', 'RSI_signal']].apply('TimeToBuy', axis=1)

-----------------------------------------------------------------------------
.diff(2) = 1

dt_df['first_crashed'] = dt_df['time_to_buy'].apply('diff')

idx = 0
idx_ls = []
fs_np = np.array(dt_df['first_crashed'])

idx = np.where(fs_np == 1.0)
idx_ls.append(idx[0]+1)

idx_ls에 있는 인덱스들의 dt컬럼에 있는 날짜가 매수일이 된다.
매수일 알아보기
purchase_date = []
for idx in idx_ls:
  purchase_date.append(dt_df.iloc[idx]['dt'])

In [ ]:
def TimeToBuy(e):
  timing = 0
  if e['RSI'] < 0.3 and e['RSI_signal'] < 0.3:
    timing = 1
  return timing

In [ ]:
dt_df['time_to_buy'] = dt_df[['RSI', 'RSI_signal']].apply(TimeToBuy, axis=1)

In [ ]:
pd.set_option('display.max_rows', None)

In [ ]:
pd.set_option('display.max_columns', None)

In [ ]:
dt_df['first_crashed'] = dt_df['time_to_buy'].apply('diff')

In [ ]:
idx = 0
idx_ls = []
fs_np = np.array(dt_df['first_crashed'])

idx = np.where(fs_np == 1.0)
idx_ls.append(idx[0]+1)

In [ ]:
idx

(array([115, 126, 142, 189, 228]),)

In [ ]:
idx_ls

[array([116, 127, 143, 190, 229])]

In [ ]:
purchase_date = []
for idx in idx_ls:
  purchase_date.append(dt_df.iloc[idx]['DT'])

In [ ]:
# 매수일 결정!
purchase_date

[116   2019-06-24
 127   2019-07-09
 143   2019-07-31
 190   2019-10-11
 229   2019-12-05
 Name: DT, dtype: datetime64[ns]]

In [ ]:
#dt_df

매도일 산정

In [ ]:
def TimeToSell(e): # e는 Ser
  timing = 0
  if e['RSI'] > 0.7 and e['RSI_signal'] > 0.7:
    timing = 1
  return timing

In [ ]:
dt_df['time_to_sell'] = dt_df[['RSI', 'RSI_signal']].apply(TimeToSell, axis=1)

In [ ]:
dt_df['first_rasing'] = dt_df['time_to_sell'].apply('diff')

In [ ]:
idx = 0
idx_ls = []
fs_np = np.array(dt_df['first_rasing'])

idx = np.where(fs_np == 1.0)
idx_ls.append(idx[0]+1)

In [ ]:
idx

(array([ 33,  66,  99, 175, 243]),)

In [ ]:
idx_ls

[array([ 34,  67, 100, 176, 244])]

In [ ]:
sale_date = []
for idx in idx_ls:
  sale_date.append(dt_df.iloc[idx]['DT'])

In [ ]:
# 매도일 결정!
sale_date

[34    2019-02-22
 67    2019-04-11
 100   2019-05-30
 176   2019-09-19
 244   2019-12-27
 Name: DT, dtype: datetime64[ns]]


매수일
116   2019-06-24 -> 매수!
127   2019-07-09
143   2019-07-31
190   2019-10-11 -> 2차 매수!
229   2019-12-05

총 매수 주식 수 :  (주)
매수를 마치고 남은 잔액 :  원

매수
6/24
원 어치 매수
116   2019-06-24

6/24 종가를 기준으로 주식수량 산정
2019/06/24 종가 : 2,045 원
약 1억원으로  48,899 주 매수가능

매수에 99,998,455 원 사용
잔액 : 1,545 원

In [ ]:
# 2019/06/24 종가 : 2,045 원
dt_df['C_PRC'][116]

2045

In [ ]:
# 약 1억원으로 48,899 주 매수가능
int(100000000 / dt_df['C_PRC'][116])

48899

In [ ]:
# 매수에 99,998,455 원 사용
# 잔액 : 1,545 원
dt_df['C_PRC'][116] * 48899

99998455

In [ ]:
# 매수 후 잔액 : 1,545 원
100000000 - 99998455

1545

매도하기

매수를 마친 잔액은 1,670 원
48,899 주 보유중

34    2019-02-22
67    2019-04-11
100   2019-05-30
176   2019-09-19 -> 매도!
244   2019-12-27 -> 2차 매도!

In [ ]:
# 매도일인 176  2019-09-19에 종가 : 1,670원
dt_df['C_PRC'][176]

1670

매도 이전 총 보유 주식 수 : 48,899 주

매도 가능한 기준이 되는 첫번째 시점에 전량 매도!
120   2019-06-28 (48,899 주 매도)
매도일에 종가 : 1,670 원
매도 수익금 = 81,661,330 원

1차 매도 후 잔액 = 81,662,875 원

In [ ]:
# 매도 수익금 = 81661330 원
dt_df['C_PRC'][176] * 48899

81661330

In [ ]:
# 순 수익금 : -18, 337,125 원
81661330 - 99998455

-18337125

In [ ]:
# 1차 매도 후 잔액 = 81,662,875 원
1545 + 81661330

81662875

In [ ]:
# 1차!!!
# BUY(산 날) 종가 : 2,045(원)
buy_c_prc = 2045

In [ ]:
# SELL(판 날) 종가 : 1,670 (원)
sell_c_prc = 1670

In [ ]:
# 1차 수익률
( (sell_c_prc - buy_c_prc) / buy_c_prc ) * 100

-18.337408312958438

# 2차 매수/매도

매수일
116   2019-06-24 -> 1차 매수
127   2019-07-09
143   2019-07-31
190   2019-10-11 -> 2차 매수!
229   2019-12-05

총 매수 주식 수 :  (주)
1차 매수를 마치고 남은 잔액 : 81662875 원

매수
10-11
원 어치 매수
190   2019-10-11

7/25 종가를 기준으로 주식수량 산정
2019/10-11 종가 : 1,460 원
잔액( 81,662,875 원)으로  55,933 주 매수가능

매수에  81,662,180 원 사용
잔액 :  695 원

In [ ]:
# 2019/10/11 종가 : 1,460 원
dt_df['C_PRC'][190]

1460

In [ ]:
# 81,662,875 원으로 55,933 주 매수가능
int(81662875 / dt_df['C_PRC'][190])

55933

In [ ]:
# 매수에 81,662,180 원 사용
# 잔액 : 1,750 원
dt_df['C_PRC'][190] * 55933

81662180

In [ ]:
# 매수 후 잔액 : 695 원
81662875 - 81662180

695

# 2차 매도하기

매수를 마친 잔액은 695 원
55,933 주 보유중

34    2019-02-22
67    2019-04-11
100   2019-05-30
176   2019-09-19 -> 1차 매도!
244   2019-12-27 -> 2차 매도!

In [ ]:
# 매도일인 2019-12-27에 종가 : 1,610원
dt_df['C_PRC'][244]

1610

매도 이전 총 보유 주식 수 : 55,933 주

매도 가능한 기준이 되는 첫번째 시점에 전량 매도!
244   2019-12-27 (55,933 주 매도)
매도일에 종가 : 1,610 원
매도 수익금 =    원

최종잔액 =   원

In [ ]:
# 매도 수익금 = 90,052,130 원
dt_df['C_PRC'][244] * 55933

90052130

In [ ]:
# 순 수익금 : 8,389,950 원
90052130 - 81662180

8389950

In [ ]:
# 최종잔액 = 90,052,825 원
695 + 90052130

90052825

In [ ]:
# 2차!!!
# BUY(산 날) 종가 : 1,460 (원)
buy_c_prc = 1460

In [ ]:
# SELL(판 날) 종가 : 1,610 (원)
sell_c_prc = 1610

In [ ]:
# 2차 수익률
( (sell_c_prc - buy_c_prc) / buy_c_prc ) * 100

10.273972602739725

In [ ]:
fig = make_subplots(
    rows=4, cols=1,
    vertical_spacing=0.01,
    shared_xaxes=True,
    subplot_titles=('<b>바른손이앤에이 봉차트</b>',
                    '<b>Select Data Range</b>',
                    '<b>RSI & RSI signal</b>')
)

# 봉차트
fig.add_trace(
    go.Candlestick(
        x = dt_df['DT'],
        open = dt_df['O_PRC'],
        high = dt_df['H_PRC'],
        low = dt_df['L_PRC'],
        close = dt_df['C_PRC'],
        increasing_line_color = 'red',
        decreasing_line_color = 'blue'
    ), row=1, col=1
)

# RSI 선 그래프
fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['RSI'],
        line = dict(color='blue'),
        name='RSI'
    ), row=3, col=1

)

# RSI signal 선 그래프
fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['RSI_signal'],
        line = dict(color='red')
        , name='RSI_signal'
    ), row=3, col=1

)

fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['S3'],
        line = dict(color='red', dash='dash')
        , name='High bound'
    ), row=3, col=1
)

fig.add_trace(
    go.Scatter(
        x = dt_df['DT'],
        y = dt_df['S7'],
        line = dict(color='blue', dash='dash')
        , name='Low bound'
    ), row=3, col=1
)

# 매수 적합시점 표시(실제 매수날 제외)
fig.add_trace(
    go.Scatter(
        x = ['2019-07-09', '2019-07-31', '2019-12-05'],
        y = [0.337278, 0.209677, 0.131868],
        mode = "markers+text",
        marker = dict(color = 'darkred', size = 10, symbol = 6),
        text = ['07-09', '07-31', '12-05'],
        textfont = dict(size = 15),
        textposition = "bottom center",
        name = 'RSI 기반 매수 적합시점'
    ), row = 3, col = 1
)

# 실제 매수 시점 표시
fig.add_trace(
    go.Scatter(
        x = ['2019-06-24', '2019-10-11'],
        y = [0.114286, 0.243902],
        mode = "markers+text",
        marker = dict(color = 'black', size = 10, symbol = 17),
        text = ['06-24', '10-11'],
        textfont = dict(size = 15),
        textposition = "bottom center",
        name = 'RSI 기반 매수시점'
    ), row = 3, col = 1
)

# 매도 적합시점 표시(실제 매도시점 제외)
fig.add_trace(
    go.Scatter(
        x = ['2019-02-22', '2019-04-11', '2019-05-30'],
        y = [0.825, 0.912621, 0.915408],
        mode = "markers+text",
        marker = dict(color = 'darkcyan', size = 10, symbol = 5),
        text = ['02-22', '04-11', '05-30'],
        textfont = dict(size = 13),
        textposition = "top center",
        name = 'RSI 기반 매도 적합시점'
    ), row = 3, col = 1
)

# 실제 매도 시점 표시
fig.add_trace(
    go.Scatter(
        x = ['2019-09-19', '2019-12-27'],
        y = [0.808824, 0.857143],
        mode = "markers+text",
        marker = dict(color = 'black', size = 10, symbol = 17),
        text = ['09-19', '12-27'],
        textfont = dict(size = 13),
        textposition = "top center",
        name = 'RSI 기반 매도시점'
    ), row = 3, col = 1
)

fig.update_layout(width = 1400, height = 1400, xaxis_rangeslider_visible = True)

fig.show()